[RDY] vim-patch:8.1.{265,271,273,274,275,277,278,279,280,281,282,284,286,291,295,296,320,321,339,351,392,399,552}

aufarg opened the pull request:
Hello, this is my first time porting patches. I made some changes from the original diff, notably:
1. Change incsearch state argument name from `is_state` to `s`
2. I'm reusing `command_line_next_incsearch` for `may_adjust_incsearch_highlighting`
3. `may_do_incsearch_highlighting` returns a value

aufarg committed:
vim-patch:8.1.0265: the getcmdline() function is way too big

Problem:    The getcmdline() function is way too big.
Solution:   Factor out the incremental search highlighting.
https://github.com/vim/vim/commit/0ee81cb63876e6ec1b2a6e0125295b43f1c63c7d

aufarg commented:
CI error seems not related to the changes

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -252,6 +258,186 @@ static void restore_viewstate(viewstate_T *vs)
 
+static void init_incsearch_state(incsearch_state_T *s)
+{
+    s->match_start = curwin->w_cursor;
```
2-space indent.

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -252,6 +258,186 @@ static void restore_viewstate(viewstate_T *vs)
+}
+
+// Return TRUE when 'incsearch' highlighting is to be done.
+static int do_incsearch_highlighting(int firstc)
```
bool

janlazo commented:
Did you make the changes based on later patches? If so, I suggest to port them over. There are many follow-up patches to the `incsearch` feature.

janlazo commented:
Ported patch looks good aside from style issues and C99 tweaks (ie. bool).

aufarg committed:
vim-patch:8.1.0271: 'incsearch' doesn't work for :s, :g or :v

Problem:    'incsearch' doesn't work for :s, :g or :v.
Solution:   Also use 'incsearch' for other commands that use a pattern.
https://github.com/vim/vim/commit/b0acacd767a2b0618a7f3c08087708f4329580d0

aufarg committed:
vim-patch:8.1.0274: 'incsearch' triggers on ":source"

Problem:    'incsearch' triggers on ":source".
Solution:   Check for the whole command name.
https://github.com/vim/vim/commit/21f990e1c22ffa2fdb66a548ebbe25e6e7194776

aufarg committed:
vim-patch:8.1.0275: 'incsearch' with :s doesn't start at cursor line

Problem:    'incsearch' with :s doesn't start at cursor line.
Solution:   Set cursor before parsing address. (closes vim/vim#3318)
            Also accept a match at the start of the first line.
https://github.com/vim/vim/commit/976b847f43dd16eb6cd809d2dcab7dde6045e176

aufarg committed:
vim-patch:8.1.0277: 'incsearch' highlighting wrong in a few cases

Problem:    'incsearch' highlighting wrong in a few cases.
Solution:   Fix using last search pattern.  Restore highlighting when changing
            command. (issue vim/vim#3321)
https://github.com/vim/vim/commit/c7f08b7ee1c1ff2080d425c2fcdb6907c26fc98e

aufarg committed:
vim-patch:8.1.0278: 'incsearch' highlighting does not accept reverse range

Problem:    'incsearch' highlighting does not accept reverse range.
Solution:   Swap the range when needed. (issue vim/vim#3321)
https://github.com/vim/vim/commit/60d0871000e9abf3716ee035cba5b5a9d659e327

aufarg committed:
vim-patch:8.1.0279: 'incsearch' highlighting does not skip white space

Problem:    'incsearch' highlighting does not skip white space.
Solution:   Skip white space after the command. (issue vim/vim#3321)
https://github.com/vim/vim/commit/2b926fcb3c5d8bd09a219009336bbec7c66ae67e

aufarg committed:
vim-patch:8.1.0273: invalid memory access when using 'incsearch'

Problem:    Invalid memory access when using 'incsearch'.
Solution:   Reset "patlen" when using previous search pattern.
https://github.com/vim/vim/commit/ef73a28401cde680dc0d02d4b0fb19ed659ec1d3

aufarg committed:
vim-patch:8.1.0280: 'incsearch' highlighting does not work for ":g!/"

Problem:    'incsearch' highlighting does not work for ":g!/".
Solution:   Skip the exclamation mark. (Hirohito Higashi)
https://github.com/vim/vim/commit/def7b1dc6104a6ce6d7c3e3a615231178601b124

aufarg commented:
I've added some later patches. I changed my mind about making `may_do_incsearch_highlighting` returns a value, instead now it runs when `inccommand` highlighting is not being done, since IIUC `inccommand` has its own highlighting.

teto commented:
Thanks @aufarg for the welcome contribution ! and kudos to @janlazo for helping you throughout, could you "Approve" the PR (via the github ui) if you think it's ready ? I will then try to run it on my machine.

janlazo commented:
@aufarg Can you add more `incsearch` patches, found by `vim-patch.sh -l` with `--grep incsearch` or `-G insearch`, up to `vim-patch:8.1.0399`? It's fine if you port only the remaining 8.1.02xx patches. I can pick up the rest. Merging your PR as is would mean that a contributor or maintainer has to immediately port the rest to avoid unexpected issues.

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -1968,6 +2147,8 @@ static int command_line_changed(CommandLineState *s)
     close_preview_windows();
     update_screen(SOME_VALID);  // Clear 'inccommand' preview.
+  } else {
+    may_do_incsearch_highlighting(s->firstc, s->count, &s->is_state);
```
Does `incsearch` interfere with `inccomand`?

aufarg replied:
They uses different kind of highlighting and kind of interfere with each other. I tried by doing this at first:
```
diff --git a/src/nvim/ex_getln.c b/src/nvim/ex_getln.c
index c58c9b172..bec12ab3b 100644
--- a/src/nvim/ex_getln.c
+++ b/src/nvim/ex_getln.c
@@ -2116,6 +2116,7 @@ static int command_line_changed(CommandLineState *s)
     }
   }
 
+  may_do_incsearch_highlighting(s->firstc, s->count, &s->is_state);
   // 'incsearch' highlighting.
   if (s->firstc == ':'
       && current_sctx.sc_sid == 0    // only if interactive
@@ -2147,8 +2148,6 @@ static int command_line_changed(CommandLineState *s)
     State = (State & ~CMDPREVIEW);
     close_preview_windows();
     update_screen(SOME_VALID);  // Clear 'inccommand' preview.
-  } else {
-    may_do_incsearch_highlighting(s->firstc, s->count, &s->is_state);
   }
 
   if (cmdmsg_rl || (p_arshape && !p_tbidi && enc_utf8)) {
```
If both run at the same time, `may_do_incsearch_highlighting` make the first match's `reverse` attribute set to true, and it fails some of the existing tests. I also tried with having a return value returned by `may_do_incsearch_highlighting`, so that only when it fails we do the `inccommand` block without success.

It also fails some other tests like `:substitute, inccommand=split deactivates if 'redrawtime' is exceeded #5602`, but I can't figure out why.

I figure that since before patch 8.1.0271 `incsearch` doesn't work with `:s`, that means whenever `inccommand` was set, we never uses the highlighting from `incsearch` on `:s` and instead we use the highlighting from `inccommand` block. So I thought we can keep current behaviour by not using `incsearch` highlighting on `:s` and use `inccommand` highlighting like we did before.

janlazo commented on src/nvim/testdir/test_search.vim:
```diff
@@ -346,25 +346,90 @@ func Test_searchc()
+  1
+  call feedkeys(":%s/the\<c-l>/xxx\<cr>", 'tx')
+  call assert_equal('  2 xxxe', getline('.'))
+
```
Missing tests from patch 8.1.0274.

aufarg replied:
For this hunk:
```
diff --git a/src/nvim/testdir/test_search.vim b/src/nvim/testdir/test_search.vim
index 71de172dc..7e8ddbe64 100644
--- a/src/nvim/testdir/test_search.vim
+++ b/src/nvim/testdir/test_search.vim
@@ -516,11 +522,11 @@ func Test_search_cmdline7()
   endif
   " need to disable char_avail,
   " so that expansion of commandline works
   call test_override("char_avail", 1)
   new
-  let @/='b'
+  let @/ = 'b'
   call setline(1, [' bbvimb', ''])
   set incsearch
   " first match
   norm! gg0
   " moves to next match of previous search pattern, just like /<cr>
/
```
It changes the line on `Test_search_cmdline7`, which was not ported from mainline. It was introduced in patch 8.0.1250. Should I port the whole test? Should it be part of this changeset or should I make another PR?

janlazo replied:
Port the whole test. It was omitted because of `test_override` but that makes it harder to port patches.

janlazo commented:
Everything else looks fine based only on the code.

aufarg commented:
> Vim doesn't accept `^T` delimiter since patch 8.1.0271 so `:%s^Tfoo^Tbar^Tg` does not work. Does that break any use case?

I'm not really sure because I just found that Vim does accept other delimiter other than `/` (The help on `:s` doesn't mention anything about the delimiter). How do I go about finding out if it break any use case?

janlazo commented:
@aufarg 

> @janlazo
> 
> > Can you add more incsearch patches, found by vim-patch.sh -l with --grep incsearch or -G insearch, up to vim-patch:8.1.0399? It's fine if you port only the remaining 8.1.02xx patches. I can pick up the rest. Merging your PR as is would mean that a contributor or maintainer has to immediately port the rest to avoid unexpected issues.
> 
> Sure, I'm fine with that, but I do have question regarding patches unrelated to `incsearch` that also needs to be ported. In fact, I stopped 8.1.0280 because 8.1.0282 introduce patch to `ex_docmd.c` which seems to miss some of previous patches. Should I also port them or not?

Can you list those missing patches?

> > Vim doesn't accept `^T` delimiter since patch 8.1.0271 so `:%s^Tfoo^Tbar^Tg` does not work. Does that break any use case?
> 
> I'm not really sure because I just found that Vim does accept other delimiter other than `/` (The help on `:s` doesn't mention anything about the delimiter). How do I go about finding out if it break any use case?

That question was meant for Neovim core team and whoever contributed to `inccommand` feature. /cc @KillTheMule 
Check Vim's github issues about incsearch or delimiter. If you find nothing, then it's probably fine.

KillTheMule commented:
> That question was meant for Neovim core team and whoever contributed to `inccommand` feature. /cc @KillTheMule

I don't have any special insight here, there's nothing in inccomand that would make ^T of special use. I guess if vim does it we should be doing it, too.

aufarg commented:
@janlazo 
>
> Can you list those missing patches?
> 

These are the missing patches for `ex_docmd.c` before v8.1.0282:
```
  • v8.0.1459: cannot handle change of directory
  • v8.0.1505: debugger can't break on a condition
  • v8.0.1570: can't use :popup for a menu in the terminal
  • v8.0.1592: terminal windows in a session are not properly restored
  • v8.0.1804: using :normal in terminal window causes problems
  • v8.1.0035: not easy to switch between prompt buffer and other windows
  • v8.1.0095: dialog for ":browse tabnew" says "new window"
  • v8.1.0149: session is wrong with multiple tabs when :lcd was used
  • v8.1.0228: dropping files is ignored while Vim is busy
  • v8.1.0281: parsing command modifiers is not separated
```

aufarg committed:
vim-patch:8.1.0281: parsing command modifiers is not separated

Problem:    Parsing command modifiers is not separated.
Solution:   Move command modifier parsing to a separate function.
https://github.com/vim/vim/commit/effed9315c6c5a35fc2824b90da4af753c7a02dc

aufarg committed:
vim-patch:8.1.0282: 'incsearch' does not work with command modifiers

Problem:    'incsearch' does not work with command modifiers.
Solution:   Skip command modifiers.
https://github.com/vim/vim/commit/33c4dbb74bdf41aadd193a704f597d4df20f0e47

aufarg committed:
vim-patch:8.1.0284: 'cursorline' highlighting wrong with 'incsearch'

Problem:    'cursorline' highlighting wrong with 'incsearch'.
Solution:   Move the cursor back if the match is outside the range.
https://github.com/vim/vim/commit/2f6a346a4cd2d5bdd6dc9b3209ebce7b6340221d

aufarg committed:
vim-patch:8.1.0291: 'incsearch' highlighting not used for :sort

Problem:    'incsearch' highlighting not used for :sort.
Solution:   Handle pattern in :sort command.
https://github.com/vim/vim/commit/81f56536b1bc324eb173924a8cf4d7dbbf4f3fdb

aufarg committed:
vim-patch:8.1.0295: no 'incsearch' highlighting for :vimgrep and similar

Problem:    No 'incsearch' highlighting for :vimgrep and similar commands.
Solution:   Parse the :vimgrep command and similar ones to locate the search
            pattern. (Hirohito Higashi, closes vim/vim#3344)
https://github.com/vim/vim/commit/264cf5cfaf40e704aea2578e70c15ed9a9d0161e

aufarg committed:
vim-patch:8.1.0320: too much 'incsearch' highlight for pat matching everything

Problem:    Too much 'incsearch' highlight for pattern matching everything.
Solution:   Add the skiplen to the command and remove the line range.
            (Christian Brabandt)  Check for empty pattern earlier.
https://github.com/vim/vim/commit/8b0d5ce881ac16a36ea00018ba13a58b0fdb7534

aufarg committed:
vim-patch:8.1.0321: 'incsearch' regression: /\v highlights everything

Problem:    'incsearch' regression: /\v highlights everything.
Solution:   Put back the empty_pattern() check.
https://github.com/vim/vim/commit/4edfe2d2a2d70ea66a7f73e9b923c2d1f6246a57

aufarg committed:
vim-patch:8.1.0339: wrong highlight when 'incsearch' set and cancelling :s

Problem:    Wrong highlight when 'incsearch' set and cancelling :s.
Solution:   Reset search line range. (Hirohito Higashi, Masamichi Abe)
https://github.com/vim/vim/commit/f13daa46da85a80dd05704cdde0660c2b2651a5a

aufarg committed:
vim-patch:8.1.0351: 'incsearch' for :/foo/s//<Esc> changes last search pattern

Problem:    'incsearch' for :/foo/s//<Esc> changes last search pattern.
Solution:   Save the last search pattern earlier.
https://github.com/vim/vim/commit/198cb66d652d3d8ac16226dcc929a11b0b720151

aufarg committed:
vim-patch:8.1.0356: using :s with 'incsearch' prevents CTRL-R CTRL-W

Problem:    Using :s with 'incsearch' prevents CTRL-R CTRL-W. (Boris Staletic)
Solution:   When past the pattern put cursor back in the start position.
            (closes vim/vim#3413)
https://github.com/vim/vim/commit/99f043a57d0be35ef72572b0429cf51525c3cd2b

aufarg committed:
vim-patch:8.1.0399: 'hlsearch' highlight remains in other window

Problem:    'hlsearch' highlight remains in other window after cancelling
            command.
Solution:   Redraw all windows. Also remove unnecessary delays. (closes vim/vim#3437)
https://github.com/vim/vim/commit/65985ac998713dfe0f0ff1dd49c5e3e8f17f4870

aufarg committed:
vim-patch:8.1.0286: 'incsearch' does not apply to :smagic and :snomagic

Problem:    'incsearch' does not apply to :smagic and :snomagic.
Solution:   Add support. (Hirohito Higashi)
https://github.com/vim/vim/commit/167ae42685dcd430800c51ac7339f7f0938a3e70

aufarg committed:
vim-patch:8.1.0296: command parsing for 'incsearch' is a bit ugly

Problem:    Command parsing for 'incsearch' is a bit ugly.
Solution:   Return when there is no pattern.  Put common checks together.
https://github.com/vim/vim/commit/111bbd61e96d1d3ee4bc07ae435fd97b88413aba

aufarg committed:
vim-patch:8.1.0392: error while typing :/foo/s// with 'incsearch' enabled

Problem:    Error while typing :/foo/s// with 'incsearch' enabled.
Solution:   Do not give search errors when highlighting matches.
https://github.com/vim/vim/commit/50eb16c3b23235b21ce4494673a7741a9a196176

aufarg committed:
ex_docmd: merge parse_state_T with exarg_T

janlazo commented:
Which patch caused the test failures?

aufarg commented:
> Which patch caused the test failures?

All of the failures seems to be caused by 8.1.0296.

aufarg commented:
I added one patch to merge `parsed_state_T` on `ex_docmd.c` to `exarg_T`. Will merge with 8.1.0281 if the patch is okay.

janlazo commented:
If you break this PR to 2 branches, one for patches before and one for patches after patch 8.1.0296 (or patch 8.1.0281 because of your patch), are there patches in the latter branch than you can cherry-pick to the former branch with minimal or no difficulty and without failing the tests? 

I can't merge the PR if there are failing tests.

justinmk commented:
re https://github.com/neovim/neovim/pull/12721#discussion_r471073908 

> I figure that since before patch 8.1.0271 `incsearch` doesn't work with `:s`, that means whenever `inccommand` was set, we never uses the highlighting from `incsearch` on `:s` and instead we use the highlighting from `inccommand` block. So I thought we can keep current behaviour by not using `incsearch` highlighting on `:s` and use `inccommand` highlighting like we did before.

Yes, that is the right approach.

aufarg commented:
When I try to test with UBSan in local, it always failed with:
```
[----------] Running tests from test/functional/legacy/search_spec.lua
-- Tests exited non-zero: SIGPIPE
-- No output to stderr.
```
Any pointers?

aufarg committed:
vim-patch:8.1.0552: saved last search pattern may not be restored

Problem:    Saved last search pattern may not be restored.
Solution:   Call restore_last_search_pattern().  Add a check for balancing
            saving and restoring the last search pattern.
https://github.com/vim/vim/commit/01a060da74b756b161e595b39a7877cb7612120a

aufarg commented:
Found the fix for the leak detected by UBSan in v8.1.0552. I've added it to the list of the patches.

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -146,6 +147,19 @@ typedef struct {
+  pos_T       match_start;
+  pos_T       match_end;
+  int         did_incsearch;
+  int         incsearch_postponed;
```
bool

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -146,6 +147,19 @@ typedef struct {
+  viewstate_T old_viewstate;
+  pos_T       match_start;
+  pos_T       match_end;
+  int         did_incsearch;
```
bool

janlazo commented on src/nvim/globals.h:
```diff
@@ -353,9 +353,11 @@ EXTERN int t_colors INIT(= 256);                // int value of T_CCO
-EXTERN int highlight_match INIT(= false);       // show search match pos
-EXTERN linenr_T search_match_lines;             // lines of of matched string
-EXTERN colnr_T search_match_endcol;             // col nr of match end
+EXTERN int highlight_match INIT(= false);          // show search match pos
```
in general, if the variable is set to true/false only, use bool always. If it is set to `MAYBE`,  use TriState. If it's set to anything else, keep int and use integer values directly even if Vim uses true,false for 1,0.

janlazo commented on src/nvim/ex_docmd.c:
```diff
@@ -1369,29 +1264,43 @@ static char_u * do_one_cmd(char_u **cmdlinep,
-  parsed.save_msg_silent = -1;
-  parsed.did_esilent = 0;
-  parsed.did_sandbox = false;
-  bool parse_success = parse_one_cmd(cmdlinep, &parsed, cstack, fgetline, cookie);
```
Is this removed because `parse_one_cmd` doesn't do as documented or is it useless or is it to make patching easier?

janlazo replied:
Did you check https://github.com/neovim/neovim/pull/11340?

aufarg replied:
I removed `parse_one_cmd` because it seems that the function does the same thing as the additional code from v8.1.0281 on `do_one_cmd`, except it brings no side-effect to the globals. The function is called only in one place though, so I wonder if making the function have no side-effect warrants the additional effort to port later patches. It also seems that the patch achieve the same goal as #11340.

That said, I'm not aware if the team have something in mind that requires the function to not do any side-effect. Let me know if I shouldn't remove it.

janlazo replied:
Will have to revist this for recent Vim 8.2 patches such as https://github.com/vim/vim/commit/5661ed6c833e05467cab33cb9b1c535e7e5cc570.

janlazo commented:
Tests added/updated in test_search.vim are not in sync with Vim.
- Test_search_cmdline3s

janlazo commented on src/nvim/testdir/test_search.vim:
```diff
@@ -520,12 +637,203 @@ func Test_incsearch_substitute_dump()
+    return
+  endif
+  if !CanRunVimInTerminal()
+    throw 'Skipped: cannot make screendumps'
```
In order to verify that your changes work, convert these screendump tests into lua functional screen tests in `test/functional/`. These codecov warnings appear because these screendump tests are skipped.

You can ignore the `test_override` tests. As far as I know, Neovim does not support modifying the internal state, at least within Vimscript.

aufarg replied:
I've added the screendump tests, not sure if I put it in the correct place though. Also the codecov warnings are still there.

janlazo commented on src/nvim/ex_getln.c:
```diff
@@ -252,6 +259,382 @@ static void restore_viewstate(viewstate_T *vs)
+  save_viewstate(&s->old_viewstate);
+}
+
+// Return TRUE when 'incsearch' highlighting is to be done.
```
true

janlazo commented on src/nvim/testdir/test_search.vim:
```diff
@@ -423,6 +502,48 @@ func Test_search_cmdline5()
+  endif
+  " Test that an pressing <c-g> in an empty command line
+  " does not move the cursor
+  if !exists('+incsearch')
```
Duplicate `+incsearch` check?

aufarg commented:
I've changed all occurence of `TRUE` on my added lines to `true` and changed `delim_optional` to be `bool`. I think that's all that I can find that should be changed from the original patch.

janlazo commented:
I'll review again and approve if it's ready. Good work!

janlazo reviewed (approved):
LGTM

janlazo commented:
@teto Are you testing this PR?

teto commented:
@janlazo no need if you approved of it ;) I'll let you merge since you have the rights now :)
